src/scanner.c

```diff
@@ -0,0 +1,593 @@
+    {                                                                          \
+        if ((vec)->data != NULL)                                               \
+            free((vec)->data);                                                 \
+    }
+
+#define VEC_CLEAR(vec)                                                         \
+    {                                                                          \
+        for (int i = 0; i < (vec)->len; i++) {                                 \
+            STRING_FREE((vec)->data[i].word);                                  \
+        }                                                                      \
+        (vec)->len = 0;                                                        \
+    }
+
+#define STRING_RESIZE(vec, _cap)                                               \
+    (vec)->data = realloc((vec)->data, (_cap + 1) * sizeof((vec)->data[0]));   \
+    assert((vec)->data != NULL);                                               \
+    memset((vec)->data + (vec)->len, 0,                                        \
+           ((_cap + 1) - (vec)->len) * sizeof((vec)->data[0]));                \
+    (vec)->cap = (_cap);
+
+#define STRING_GROW(vec, _cap)                                                 \
+    if ((vec)->cap < (_cap)) {                                                 \
+        STRING_RESIZE((vec), (_cap));                                          \
+    }
+
+#define STRING_PUSH(vec, el)                                                   \
+    if ((vec)->cap == (vec)->len) {                                            \
+        STRING_RESIZE((vec), MAX(16, (vec)->len * 2));                         \
+    }                                                                          \
+    (vec)->data[(vec)->len++] = (el);
+
+#define STRING_FREE(vec)                                                       \
+    {                                                                          \
+        if ((vec)->data != NULL)                                               \
+            free((vec)->data);                                                 \
+        free((vec));                                                           \
+    }
+
+enum TokenType {
+    AUTOMATIC_SEMICOLON,
+    ENCAPSED_STRING_CHARS,
+    ENCAPSED_STRING_CHARS_AFTER_VARIABLE,
+    EXECUTION_STRING_CHARS,
+    EXECUTION_STRING_CHARS_AFTER_VARIABLE,
+    ENCAPSED_STRING_CHARS_HEREDOC,
+    ENCAPSED_STRING_CHARS_AFTER_VARIABLE_HEREDOC,
+    EOF_TOKEN,
+    HEREDOC_START,
+    HEREDOC_END,
+    NOWDOC_STRING,
+    SENTINEL_ERROR, // Unused token used to indicate error recovery mode
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    char *data;
+} String;
+
+String *string_new() {
+    String *str = calloc(1, sizeof(String));
+    str->data = calloc(1, sizeof(char) * 16);
+    return str;
+}
+
+typedef struct {
+    String *word;
```

Comment:
Also, can this string be stored inline, instead of allocating it separately?

Reply:
done